pkg/syncer/resourcesync/controller.go

```diff
@@ -0,0 +1,330 @@
+			if err != nil {
+				return false
+			}
+			clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+			if err != nil {
+				return false
+			}
+			return name == syncTargetName && clusterName == syncTargetWorkspace
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) { c.enqueueSyncTarget(obj, logger) },
+			UpdateFunc: func(old, obj interface{}) {
+				oldCluster := old.(*workloadv1alpha1.SyncTarget)
+				newCluster := obj.(*workloadv1alpha1.SyncTarget)
+
+				// only enqueue when syncedResource is changed.
+				if !equality.Semantic.DeepEqual(oldCluster.Status.SyncedResources, newCluster.Status.SyncedResources) {
+					c.enqueueSyncTarget(obj, logger)
+				}
+			},
+			DeleteFunc: func(obj interface{}) { c.enqueueSyncTarget(obj, logger) },
+		},
+	})
+
+	return c, nil
+}
+
+func (c *Controller) enqueueSyncTarget(obj interface{}, logger logr.Logger) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logging.WithQueueKey(logger, key).V(2).Info("queueing SyncTarget")
+
+	c.queue.Add(key)
+}
+
+// Start starts the controller workers.
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), controllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) AddUpstreamEventHandler(handler ResourceEventHandlerPerGVR) {
+	c.upstreamEventHandlers = append(c.upstreamEventHandlers, handler)
+}
+
+func (c *Controller) AddDownstreamEventHandler(handler ResourceEventHandlerPerGVR) {
+	c.downstreamEventHandlers = append(c.downstreamEventHandlers, handler)
+}
+
+func (c *Controller) InformerForResource(gvr schema.GroupVersionResource) (*SyncerInformer, bool) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	if informer, ok := c.syncerInformerMap[gvr]; ok {
+		return informer, true
+	}
+
+	return nil, false
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("failed to sync %q: %w", key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	syncTarget, err := c.syncTargetLister.Get(key)
```

Comment:
This will be the wrong key format - see other syncer code for the hard-coded reformat before we get new listers in.

Reply:
hrm, I am bit confused here...I am using the same pattern as here https://github.com/kcp-dev/kcp/blob/main/pkg/reconciler/workload/synctargetexports/synctargetexports_controller.go#L265

Reply:
```go
	// TODO(skuznets): can we figure out how to not leak this detail up to this code?
	// I guess once the indexer is using kcpcache.MetaClusterNamespaceKeyFunc, we can just use that formatter ...
	var indexKey string
	if upstreamNamespace != "" {
		indexKey += upstreamNamespace + "/"
	}
	if !clusterName.Empty() {
		indexKey += clusterName.String() + "|"
	}
	indexKey += name
	// get the upstream object
	obj, exists, err := c.upstreamInformers.ForResource(gvr).Informer().GetIndexer().GetByKey(indexKey)
```

From `spec_process.go` is the correct incantation.

Reply:
Please feel free to change synctargetexports as well